src/mds/CInode.cc

```diff
@@ -5133,6 +5135,151 @@ void CInode::maybe_export_pin(bool update)
   }
 }
 
+class C_Inode_Ephemeral_Distributed_Update : public MDSLogContextBase {
+protected:
+  CInode *in; 
+  MDSRank *get_mds() override {return in->mdcache->mds;}
+  void finish(int r) override {
+    in->is_export_ephemeral_distributed = true;
+  }    
+
+public:
+  C_Inode_Ephemeral_Distributed_Update(CInode *i) : in(i) {}
+};
+
+class C_Inode_Ephemeral_Random_Update : public MDSLogContextBase {
+protected:
+  CInode *in;
+  MDSRank *get_mds() override {return in->mdcache->mds;}
+  void finish(int r) override {
+    in->is_export_ephemeral_random = true;
+  }
+
+public:
+  C_Inode_Ephemeral_Random_Update(CInode *i) : in(i) {}
+};
+
+void CInode::maybe_export_ephemeral_pin(bool update)
+{
+  //If both the configs aren't set then return
+  if (!g_conf().get_val<bool>("mds_export_ephemeral_random")
+      && !g_conf().get_val<bool>("mds_export_ephemeral_distributed"))
+    return;
+
+  double export_ephemeral_random_pin = get_export_ephemeral_random_pin(false);
+  bool export_ephemeral_distributed_pin = get_export_ephemeral_distributed_pin();
+
+
+  //Check if it's already ephemerally pinned
+  if ((is_export_ephemeral_random || is_export_ephemeral_distributed) && !update)
+      return;
+
+  if (g_conf().get_val<bool>("mds_export_ephemeral_random")) {
+    if ((export_ephemeral_random_pin >=
+	  ceph::util::generate_random_number(0.0, 1.0) || update)
+	&& export_ephemeral_distributed_pin == false) {
+
+      MDLog *mdlog = mdcache->mds->mdlog;
+
+      CDentry *pdn = get_projected_parent_dn();
+
+      EUpdate *le = new EUpdate(mdlog, "export ephemeral random");
+      mdlog->start_entry(le);
+
+      le->metablob.add_primary_dentry(pdn, this, false, false, false,
+          false, true);
+
+      mdlog->submit_entry(le, new C_Inode_Ephemeral_Random_Update(this));
+
+      bool queue = false;
+      for (auto p = dirfrags.begin(); p != dirfrags.end(); p++) {
+        CDir *dir = p->second;
+        if (!dir->is_auth())
+          continue;
+        if (dir->is_subtree_root()) {
+          // set auxsubtree bit or export it
+          if (!dir->state_test(CDir::STATE_AUXSUBTREE) ||
+	      mdcache->hash_into_rank_bucket(ino(), mdcache->mds->mdsmap->get_max_mds()) != dir->get_dir_auth().first) {
```

Comment:
Why is this not using the directory inode's ino number? i.e. why `ino()`?

Reply:
ino() will return the inode number right?(https://github.com/ceph/ceph/blob/master/src/mds/CInode.h#L598)

Reply:
Ah, I see. I mistook this for an empty constructor.